pkg/iter/iterator.go

```diff
@@ -133,38 +135,67 @@ func (i *heapIterator) requeue(ei EntryIterator) {
 }
 
 func (i *heapIterator) Next() bool {
-	if i.curr != nil {
-		i.requeue(i.curr)
-	}
-
 	if i.heap.Len() == 0 {
 		return false
 	}
 
-	i.curr = heap.Pop(i.heap).(EntryIterator)
-	currEntry := i.curr.Entry()
-
-	// keep popping entries off if they match, to dedupe
+	// We support multiple entries with the same timestamp, and we want to
+	// preserve their original order. We look at all the top entries in the
+	// heap with the same timestamp, and pop the ones whose common value
+	// occurs most often.
+	type tuple struct {
+		logproto.Entry
+		EntryIterator
+	}
+	tuples := make([]tuple, 0, i.heap.Len())
 	for i.heap.Len() > 0 {
 		next := i.heap.Peek()
-		nextEntry := next.Entry()
-		if !currEntry.Equal(nextEntry) {
+		entry := next.Entry()
+		if len(tuples) > 0 && !tuples[0].Timestamp.Equal(entry.Timestamp) {
 			break
 		}
 
-		next = heap.Pop(i.heap).(EntryIterator)
-		i.requeue(next)
+		heap.Pop(i.heap)
+		tuples = append(tuples, tuple{
+			Entry:         entry,
+			EntryIterator: next,
+		})
+	}
+
+	// Find in entry which occurs most often which, due to quorum based
+	// replication, is guaranteed to be the correct next entry.
+	sort.Slice(tuples, func(i, j int) bool {
+		return tuples[i].Line < tuples[j].Line
+	})
+	i.currEntry = tuples[0].Entry
+	count, max := 1, 1
+	for j := 1; j < len(tuples); j++ {
+		if tuples[j].Equal(tuples[j-1]) {
```

Comment:
The Equals() method of the logproto.Entry does a type check.  So we need to explicitly pass a logproto.Entry and not a `tuple`
```suggestion
		if tuples[j].Entry.Equal(tuples[j-1].Entry) {
```

Reply:
or even better, just use `tuples[j].Line == tuples[j-1].Line` as we already know the timestamps are the same.